presto-hive/src/main/java/com/facebook/presto/hive/rule/BaseHiveFilterPushdown.java

```diff
@@ -0,0 +1,681 @@
+
+        // Extract deterministic conjuncts that apply to partition columns and specify these as Constraint#predicate
+        constraint = extractDeterministicConjuncts(
+                session,
+                decomposedFilter,
+                columnHandles,
+                entireColumnDomain,
+                constraint);
+
+        HivePartitionResult hivePartitionResult = getHivePartitionResult(metadata, tableHandle, constraint, session);
+
+        TupleDomain<ColumnHandle> unenforcedConstraint = getUnenforcedConstraint(hivePartitionResult, constraint);
+
+        TupleDomain<Subfield> domainPredicate = getDomainPredicate(decomposedFilter, unenforcedConstraint);
+
+        RemainingExpressions remainingExpressions = getRemainingExpressions(tableHandle, decomposedFilter, columnHandles);
+
+        Set<String> predicateColumnNames = getPredicateColumnNames(optimizedRemainingExpression, domainPredicate);
+
+        return getConnectorPushdownFilterResult(
+                columnHandles,
+                metadata,
+                session,
+                domainPredicate,
+                remainingExpressions,
+                predicateColumnNames,
+                currentLayoutHandle,
+                tableHandle,
+                hivePartitionResult);
+    }
+
+    protected abstract ConnectorPushdownFilterResult getConnectorPushdownFilterResult(
+            Map<String, ColumnHandle> columnHandles,
+            ConnectorMetadata metadata,
+            ConnectorSession session,
+            TupleDomain<Subfield> domainPredicate,
+            RemainingExpressions remainingExpressions,
+            Set<String> predicateColumnNames,
+            Optional<ConnectorTableLayoutHandle> currentLayoutHandle,
+            ConnectorTableHandle tableHandle,
+            HivePartitionResult hivePartitionResult);
+
+    protected abstract TupleDomain<ColumnHandle> getUnenforcedConstraint(
+            HivePartitionResult hivePartitionResult,
+            Constraint<ColumnHandle> constraint);
+
+    protected abstract HivePartitionResult getHivePartitionResult(
+            ConnectorMetadata metadata,
+            ConnectorTableHandle tableHandle,
+            Constraint<ColumnHandle> constraint,
+            ConnectorSession session);
+
+    protected abstract TupleDomain<ColumnHandle> intersectionWithPartitionColumnPredicate(
+            ConnectorTableLayoutHandle currentLayoutHandle,
+            TupleDomain<ColumnHandle> entireColumnDomain);
+
+    protected abstract boolean isPushdownFilterSupported(ConnectorSession session, TableHandle tableHandle);
+
+    protected abstract ConnectorMetadata getMetadata(TableHandle tableHandle);
+
+    protected abstract PrestoWarning getPrestoWarning(TableScanNode tableScan);
+
+    protected abstract Subfield toSubfield(ColumnHandle columnHandle);
+
+    protected class Rewriter
+            extends ConnectorPlanRewriter<Void>
```

Comment:
It's pretty hard for me to understand under what circumstances and how one overrides these protected methods.  Can we think about further refactoring to make the interface footprint cleaner and smaller?

Reply:
@tdcmeehan I agree this is not ideal. These protected methods were due to some nuance between Hive and Iceberg data structures, e.g. TableHandle, ColumnHandle, ConnectorMetadata, TransactionManager etc. For example, toSubfield(ColumnHandle columnHandle) needs to get the columnHandle's name, but in HIveColumnHandle, the name is a plain string, while in IcebergColumnHandle, the name is part of Iceberg ColumnIdentity:
```
public class HiveColumnHandle
        implements ColumnHandle
{
  ...
  private final String name;
}

public class IcebergColumnHandle
        implements ColumnHandle
{
    private final ColumnIdentity columnIdentity;
...
}
public class ColumnIdentity
{
    private final int id;
    private final String name;
    private final TypeCategory typeCategory;
    private final List<ColumnIdentity> children;
```
Also the `getName()` function is not part of `ColumnHandle` interface, and therefore had to be called on different classes. There are some other differences between HiveColumnHandle and IcebergColumnHandle,  but there're also many parts quite alike. 

As another example, `getMetadata(TableHandle tableHandle)` content is almost the same but the member `transactionManager` is different. HiveTransactionManager and IcebergTransactionManager do not have common ancestors, but the implementations are quite alike.  

HiveTableLayoutHandle contains a lot more contents than IcebergTableLayout, but also lots of duplicate code. The major reason was the HiveColumnHandle and IcebergColumnHandle differences.

I think a cleaner and more systematic way is to unify the interfaces of these structures, either through extracting common ancestors, or making the Iceberg structures extend their Hive counter parts. Remember, Velox only has one Hive connector and unifying these structures would make the code much easier. Furthermore, if we need to implement Hudi or Delta table in Prestissimo, there will be similar work need to be done. How do you like this idea?